### ISSUE TITLE
DDP (2GPU) performance is almost same as single GPU with official example dataset+GraphSAGE

### ISSUE THREAD

**Issue by rewonderful** [dmlc/dgl#3609]
When I test my graph  with ddp ,no matter the hetero-graph (about 40million edges) or  the homo-graph (about 120million edges) on GraphSAGE , it looks like the ddp didn't outstanding much than single gpu training.

Thus I tried the official ddp example in  "dgl/examples/pytorch/graphsage/train_sampling_multi_gpu.py" with just a little  codes change for time consuming statistics. And the result astonished me that the official  ddp example didn't performs better as well.

This is my naivce test result.
### model part :0.1412s/step,total:21.7462
- slice_cost:**0.0269**s/step,total:4.1359
- feats2device_cost:**0.1035**s/step,total:15.9343
- blks2device:0.0037s/step,total:0.5678
- forward_cost:0.0042s/step,total:0.6488
- loss_cost:0.0001s/step,total:0.0227
- zerograd_cost:0.0002s/step,total:0.0253
- backward_cost:**0.0018**s/step,total:0.2847
- optimizerstep_cost:0.0011s/step,total:step,total:0.1705
- optimizerstep_cost:0.0008s/step,total:0.1258
		
### model_part:0.2400s/step,total:18.4815
- slice_cost:**0.0755**s/step,total:5.8157
- feats2device:**0.1484**s/step,total:11.4300
- blks2device:0.0050s/step,total:0.3875
- forward_cost:0.0064s/step,total:0.4950
- loss_cost:0.0003s/step,total:0.0195
- zerograd_cost:0.0004s/step,total:0.0289
- backward_cost:**0.0029**s/step,total:0.2259
- optimizerstep_cost:0.0010s/step,total:0.0784

It seems that in 2gpu ddp each **_sample_** get  slower 2x than single-gpu, other model compute part also get slower performance in each step
### My configuration
>All training params use default args, except set epochs=5, num-workers=10,  without evaluation
dgl-verision=0.7.2
torch-version=1.70
python:3.7.4
cuda:10.1.243
gpu:V100
platform: docker+Ubuntu 18.04.3 LTS

## BTW
Because my own graph is pretty large which cannot be fully send to GPU, thus all my test is based on sampling graph on cpu first then send it to GPU, and the test result shows 'blks2device'  cost just little time.

I've tried turn on 'pin_memory=True', ddp performs even worse.

## Here is my  test code.
```python3
import dgl
import torch as th
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from collections import OrderedDict

import dgl.multiprocessing as mp
import dgl.nn.pytorch as dglnn
import time
import argparse
from torch.nn.parallel import DistributedDataParallel
import sys
import logging
_time_format = '%m/%d/%Y, %H:%M:%S'
log_level = logging.INFO
fmt = '[%(asctime)s] [%(levelname)s] [%(name)s] %(message)s'
formatter = logging.Formatter(fmt, _time_format)
handler = logging.StreamHandler(sys.stdout)
handler.setFormatter(formatter)
logger = logging.getLogger('sage')
logger.addHandler(handler)
logger.setLevel(logging.INFO)



class SAGE(nn.Module):
    def __init__(self, in_feats, n_hidden, n_classes, n_layers, activation, dropout):
        super().__init__()
        self.init(in_feats, n_hidden, n_classes, n_layers, activation, dropout)

    def init(self, in_feats, n_hidden, n_classes, n_layers, activation, dropout):
        self.n_layers = n_layers
        self.n_hidden = n_hidden
        self.n_classes = n_classes
        self.layers = nn.ModuleList()
        if n_layers > 1:
            self.layers.append(dglnn.SAGEConv(in_feats, n_hidden, 'mean'))
            for i in range(1, n_layers - 1):
                self.layers.append(dglnn.SAGEConv(n_hidden, n_hidden, 'mean'))
            self.layers.append(dglnn.SAGEConv(n_hidden, n_classes, 'mean'))
        else:
            self.layers.append(dglnn.SAGEConv(in_feats, n_classes, 'mean'))
        self.dropout = nn.Dropout(dropout)
        self.activation = activation

    def forward(self, blocks, x):
        h = x
        for l, (layer, block) in enumerate(zip(self.layers, blocks)):
            h = layer(block, h)
            if l != len(self.layers) - 1:
                h = self.activation(h)
                h = self.dropout(h)
        return h

class AverageMeterGroup:
    """
    Average meter group for multiple average meters.
    """

    def __init__(self):
        self.meters = OrderedDict()

    def update(self, data):
        """
        Update the meter group with a dict of metrics.
        Non-exist average meters will be automatically created.
        """
        for k, v in data.items():
            if k not in self.meters:
                self.meters[k] = AverageMeter(k, ":4f")
            self.meters[k].update(v)

    def __getattr__(self, item):
        return self.meters[item]

    def __getitem__(self, item):
        return self.meters[item]

    def __str__(self):
        return "  ".join(str(v) for v in self.meters.values())

    def summary(self):
        """
        Return a summary string of group data.
        """
        return "  ".join(v.summary() for v in self.meters.values())


class AverageMeter:
    """
    Computes and stores the average and current value.

    Parameters
    ----------
    name : str
        Name to display.
    fmt : str
        Format string to print the values.
    """

    def __init__(self, name, fmt=':f'):
        self.name = name
        self.fmt = fmt
        self.reset()

    def reset(self):
        """
        Reset the meter.
        """
        self.val = 0
        self.avg = 0
        self.sum = 0
        self.count = 0

    def update(self, val, n=1):
        """
        Update with value and weight.

        Parameters
        ----------
        val : float or int
            The new value to be accounted in.
        n : int
            The weight of the new value.
        """
        self.val = val
        self.sum += val * n
        self.count += n
        self.avg = self.sum / self.count

    def __str__(self):
        fmtstr = '{name} {val' + self.fmt + '} ({avg' + self.fmt + '})'
        return fmtstr.format(**self.__dict__)

    def summary(self):
        fmtstr = '{name}: {avg' + self.fmt + '}'
        return fmtstr.format(**self.__dict__)
def load_reddit():
    from dgl.data import RedditDataset

    # load reddit data
    data = RedditDataset(self_loop=True)
    g = data[0]
    g.ndata['features'] = g.ndata['feat']
    g.ndata['labels'] = g.ndata['label']
    return g, data.num_classes


def inductive_split(g):
    """Split the graph into training graph, validation graph, and test graph by training
    and validation masks.  Suitable for inductive models."""
    train_g = g.subgraph(g.ndata['train_mask'])
    val_g = g.subgraph(g.ndata['train_mask'] | g.ndata['val_mask'])
    test_g = g
    return train_g, val_g, test_g



def compute_acc(pred, labels):
    """
    Compute the accuracy of prediction given the labels.
    """
    return (th.argmax(pred, dim=1) == labels).float().sum() / len(pred)


def evaluate(model, g, nfeat, labels, val_nid, device):
    """
    Evaluate the model on the validation set specified by ``val_nid``.
    g : The entire graph.
    inputs : The features of all the nodes.
    labels : The labels of all the nodes.
    val_nid : A node ID tensor indicating which nodes do we actually compute the accuracy for.
    device : The GPU device to evaluate on.
    """
    model.eval()
    with th.no_grad():
        pred = model.inference(g, nfeat, device, args.batch_size, args.num_workers)
    model.train()
    return compute_acc(pred[val_nid], labels[val_nid])


# def load_subtensor(nfeat, labels, seeds, input_nodes, dev_id):
#     """
#     Extracts features and labels for a subset of nodes.
#     """
#     batch_inputs = nfeat[input_nodes].to(dev_id)
#     batch_labels = labels[seeds].to(dev_id)
#     return batch_inputs, batch_labels


#### Entry point

def run(proc_id, n_gpus, args, devices, data):
    # Start up distributed training, if enabled.
    dev_id = devices[proc_id]
    if n_gpus > 1:
        dist_init_method = 'tcp://{master_ip}:{master_port}'.format(
            master_ip='127.0.0.1', master_port='12345')
        world_size = n_gpus
        th.distributed.init_process_group(backend="nccl",
                                          init_method=dist_init_method,
                                          world_size=world_size,
                                          rank=proc_id)
    logger = logging.getLogger("sage.proc:{}".format(proc_id))
    th.cuda.set_device(dev_id)

    # Unpack data
    n_classes, train_g, val_g, test_g = data

    if args.inductive:
        train_nfeat = train_g.ndata.pop('features')
        val_nfeat = val_g.ndata.pop('features')
        test_nfeat = test_g.ndata.pop('features')
        train_labels = train_g.ndata.pop('labels')
        val_labels = val_g.ndata.pop('labels')
        test_labels = test_g.ndata.pop('labels')
    else:
        train_nfeat = val_nfeat = test_nfeat = g.ndata.pop('features')
        train_labels = val_labels = test_labels = g.ndata.pop('labels')

    # if not args.data_cpu: #always false
    #     train_nfeat = train_nfeat.to(dev_id)
    #     train_labels = train_labels.to(dev_id)


    in_feats = train_nfeat.shape[1]

    train_mask = train_g.ndata['train_mask']
    val_mask = val_g.ndata['val_mask']
    test_mask = ~(test_g.ndata['train_mask'] | test_g.ndata['val_mask'])
    train_nid = train_mask.nonzero().squeeze()
    val_nid = val_mask.nonzero().squeeze()
    test_nid = test_mask.nonzero().squeeze()

    # Create PyTorch DataLoader for constructing blocks
    sampler = dgl.dataloading.MultiLayerNeighborSampler(
        [int(fanout) for fanout in args.fan_out.split(',')])
    dataloader = dgl.dataloading.NodeDataLoader(
        train_g,
        train_nid,
        sampler,
        use_ddp=n_gpus > 1,
        device=dev_id if args.num_workers == 0 else None,
        batch_size=args.batch_size,
        shuffle=False,
        drop_last=False,
        pin_memory=args.pin_memory,
        num_workers=args.num_workers)

    # Define model and optimizer
    model = SAGE(in_feats, args.num_hidden, n_classes, args.num_layers, F.relu, args.dropout)
    model = model.to(dev_id)
    if n_gpus > 1:
        model = DistributedDataParallel(model, device_ids=[dev_id], output_device=dev_id)
    loss_fcn = nn.CrossEntropyLoss()
    optimizer = optim.Adam(model.parameters(), lr=args.lr)

    # Training loop
    avg = 0
    iter_tput = []
    statis = AverageMeterGroup()

    for epoch in range(args.num_epochs):
        if n_gpus > 1:
            dataloader.set_epoch(epoch)
        tic_epoch = time.time()

        # Loop over the dataloader to sample the computation dependency graph as a list of
        # blocks.
        tic_step = time.time()
        for step, (input_nodes, seeds, blocks) in enumerate(dataloader):
            sample_cost = time.time() - tic_step
            sample_speeds = len(seeds) * n_gpus / (time.time() - tic_step)
            tic_model = time.time()

            # Load the input features as well as output labels
            # batch_inputs, batch_labels = load_subtensor(train_nfeat, train_labels,
            #                                             seeds, input_nodes, dev_id)
            tic_slice = time.time()
            batch_inputs = train_nfeat[input_nodes]
            batch_labels = train_labels[seeds]
            slice_cost = time.time() - tic_slice

            tic_feats2device = time.time()
            batch_inputs = batch_inputs.to(dev_id)
            batch_labels = batch_labels.to(dev_id)
            feats2device_cost = time.time() - tic_feats2device



            tic_blks2device = time.time()
            blocks = [block.int().to(dev_id) for block in blocks]
            blks2device = time.time() - tic_blks2device
            # Compute loss and prediction

            tic_forward = time.time()
            batch_pred = model(blocks, batch_inputs)
            forward_cost = time.time() - tic_forward

            tic_loss = time.time()
            loss = loss_fcn(batch_pred, batch_labels)
            loss_cost = time.time() - tic_loss

            tic_zerograd = time.time()
            optimizer.zero_grad()
            zerograd_cost = time.time() - tic_zerograd

            tic_backward = time.time()
            loss.backward()
            backward_cost = time.time() - tic_backward

            tic_optimizerstep = time.time()
            optimizer.step()
            optimizerstep_cost = time.time() - tic_optimizerstep

            model_cost = time.time() - tic_model

            iter_tput = len(seeds) / (time.time() - tic_step)
            step_cost = time.time() - tic_step
            statis.update({
                "step_cost": step_cost,
                "model_cost": model_cost,
                "sample_cost": sample_cost,
                "sample_speeds": sample_speeds,
                "iter_tput": iter_tput,
                "slice_cost": slice_cost,
                "feats2device_cost":feats2device_cost,
                "blks2device": blks2device,
                "forward_cost": forward_cost,
                "loss_cost": loss_cost,
                "zerograd_cost": zerograd_cost,
                "backward_cost": backward_cost,
                "optimizerstep_cost": optimizerstep_cost,
            })
            if  proc_id == 0:

                total_step = len(dataloader)
                log_str = "proc:{} |TRAIN epoch:{} | Step {:05d}/{:05d} ".format(proc_id, epoch, step, total_step, )

                for key in statis.meters:
                    if key in ["sample_speeds", "iter_tput"]:
                        log_str += "| {}:{:.4f}".format(key, statis[key].avg, )
                    else:
                        log_str += "| {}:{:.4f}/{:.4f}".format(key, statis[key].avg, statis[key].avg * total_step)
                logger.info(log_str)
            tic_step = time.time()

        if n_gpus > 1:
            th.distributed.barrier()

        epoch_cost = time.time() - tic_epoch
        statis.update({
            "epoch_cost":epoch_cost,
        })

        if proc_id == 0:
            logger.info('Epoch Time(s):avg {:.4f}'.format(statis["epoch_cost"].avg))

    if n_gpus > 1:
        th.distributed.barrier()
    if proc_id == 0:
        logger.info('Epoch Time(s):avg {:.4f}'.format(statis["epoch_cost"].avg))


if __name__ == '__main__':
    argparser = argparse.ArgumentParser("multi-gpu training")
    argparser.add_argument('--gpu-no', type=str, default='0',
                           help="Comma separated list of GPU device IDs.")
    argparser.add_argument('--dataset', type=str, default='reddit')
    argparser.add_argument('--num-epochs', type=int, default=20)
    argparser.add_argument('--num-hidden', type=int, default=16)
    argparser.add_argument('--num-layers', type=int, default=2)
    argparser.add_argument('--fan-out', type=str, default='10,25')
    argparser.add_argument('--batch-size', type=int, default=1000)
    argparser.add_argument('--log-every', type=int, default=20)
    argparser.add_argument('--eval-every', type=int, default=5)
    argparser.add_argument('--lr', type=float, default=0.003)
    argparser.add_argument('--dropout', type=float, default=0.5)
    argparser.add_argument('--pin-memory', action='store_true',)
    argparser.add_argument('--num-workers', type=int, default=0,
                           help="Number of sampling processes. Use 0 for no extra process.")
    argparser.add_argument('--inductive', action='store_true',
                           help="Inductive learning setting")
    argparser.add_argument('--data-cpu', action='store_true',
                           help="By default the script puts all node features and labels "
                                "on GPU when using it to save time for data copy. This may "
                                "be undesired if they cannot fit in GPU memory at once. "
                                "This flag disables that.")
    args = argparser.parse_args()

    devices = list(map(int, args.gpu_no.split(',')))
    n_gpus = len(devices)

    g, n_classes = load_reddit()

    # Construct graph
    g = dgl.as_heterograph(g)

    if args.inductive:
        train_g, val_g, test_g = inductive_split(g)
    else:
        train_g = val_g = test_g = g

    # Create csr/coo/csc formats before launching training processes with multi-gpu.
    # This avoids creating certain formats in each sub-process, which saves momory and CPU.
    train_g.create_formats_()
    val_g.create_formats_()
    test_g.create_formats_()
    # Pack data
    data = n_classes, train_g, val_g, test_g

    if n_gpus == 1:
        run(0, n_gpus, args, devices, data)
    else:
        procs = []
        for proc_id in range(n_gpus):
            p = mp.Process(target=run, args=(proc_id, n_gpus, args, devices, data))
            p.start()
            procs.append(p)
        for p in procs:
            p.join()


```

**Comment by yzh119** [comment]
@rewonderful could you force synchronous computation by `CUDA_LAUNCH_BLOCKING=1`? Otherwise the profiling results is not accurate.

**Comment by rewonderful** [comment]
> @rewonderful could you force synchronous computation by `CUDA_LAUNCH_BLOCKING=1`? Otherwise the profiling results is not accurate.

Thanks for reminding me the cuda synchronous feature.
After correct time measurement method,  the result still shows ddp didn't performs well, general result is showed below.

<img width="979" alt="testrst" src="https://user-images.githubusercontent.com/17673536/147525884-9ba17dea-f63b-4d21-b38e-12282b153305.png">

I realized that the cpu num may has strong affect in 'slice' operation , thus I  set OMP=1, the result shows it did affect slice stage, but normally we would not  do that, after all multithreading is a good boy. It is a pity that in my test cases this is the only situation  ddp performance as expected.
<img width="494" alt="omp=1" src="https://user-images.githubusercontent.com/17673536/147526345-83614d14-ccf9-47f6-9b1a-e12657f225a4.png">
I also test the case that two trainer process load graph independently to avoid two subprocess reading graph or tensors from  one shared parent memory(actually original fork style is what I need, my own graph is large).The result shows it's not differ much than first general test.
<img width="845" alt="spawn" src="https://user-images.githubusercontent.com/17673536/147537144-3ef37090-c8c8-43bc-abf3-e3c12dec409a.png">

In conclusion  I'm curious about:
1. Why ddp sample  stage performs bad in each step ,the 'sample' stage obviously getting worse no matter how the setting is,2x slow than single mode, this is wired, cause  dist_sampler has just a little synchronization like  'set_epoch' then reset the seeds. 
2. OMP setting will influence the performance but we usually would not restrict it to one, how to let ddp really work when omp >1.
3. The final aim, let the ddp performance as it should be.


The attachment "ddp-test.xlsx"  contains all my test result.
[ddp-test.xlsx](https://github.com/dmlc/dgl/files/7783294/ddp-test.xlsx)

**Comment by rewonderful** [comment]
@yzh119 
I tried reset the OMP setting with  OMP_NUM_THREAEDS = {docker cpu core num} and the ddp performance turned to be expected, as it is seems to be caused by multi threading issues like  graph message operation.
But it still puzzled me that in my code the whole blocks has been send to GPU before model fwd/bwd computation, why a CPU environ setting (OMP_NUM_THREAEDS)  would influence the compute speed in GPU ?

**Comment by jermainewang** [comment]
This looks like a misconfiguration that we should fix. For CPU sampling with multiple worker processes, we should always set OMP_NUM_THREADS to be 1 to avoid such confusion.

**Comment by jermainewang** [comment]
#3665 has fixed the problem by calling `torch.set_num_threads(1)` when conducting sampling on CPU.
